### PR TITLE
ARROW-9 BSON subdocument support

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: bindings/python/pymongoarrow/lib.pyx [hunk 1]
```diff
@@ -103,72 +147,90 @@ def process_bson_stream(bson_stream, context):
             while bson_iter_next(&doc_iter):
                 key = bson_iter_key(&doc_iter)
                 builder = builder_map.get(key)
+                if builder is None:
+                    builder = builder_map.get(key.decode('utf8'))
                 if builder is None and context.schema is None:
-                    # Only run if there is no schema.
-                    ftype = bson_iter_type(&doc_iter)
-                    if ftype not in _builder_type_map:
+                    """Get the appropriate builder for a document field."""
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:153]: Done

File: bindings/python/pymongoarrow/lib.pyx [hunk 2]
```diff
@@ -103,72 +147,90 @@ def process_bson_stream(bson_stream, context):
             while bson_iter_next(&doc_iter):
                 key = bson_iter_key(&doc_iter)
                 builder = builder_map.get(key)
+                if builder is None:
+                    builder = builder_map.get(key.decode('utf8'))
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:151]: Switch to always str

File: bindings/python/pymongoarrow/lib.pyx [hunk 3]
```diff
@@ -54,39 +57,80 @@ cdef const bson_t* bson_reader_read_safe(bson_reader_t* stream_reader) except? N
...
+    BSON_TYPE_BOOL: bool_(),
+    BSON_TYPE_DECIMAL128: Decimal128StringType()
+}
+
+
+cdef extract_document_dtype(bson_iter_t * doc_iter, context):
+    """Get the appropropriate data type for a sub document"""
+    cdef const char* key
+    cdef bson_type_t value_t
+    cdef bson_iter_t * child_iter
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:86]: Switched to using the stack.  Yes, Cython does do some magic when copying pointers.

File: bindings/python/pymongoarrow/lib.pyx [hunk 4]
```diff
@@ -397,3 +459,87 @@ cdef class BoolBuilder(_ArrayBuilderBase):
...
+        self.builder.get().AppendNull()
+
+    def __len__(self):
+        return self.builder.get().length()
+
+    cpdef append(self, value):
+        if not isinstance(value, bytes):
+            value = bson.encode(value)
+        process_bson_stream(value, self.context)
+        self.builder.get().Append(True)
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:536]: Added comments

```python
# Populate the child builders.
process_bson_stream(value, self.context)
# Append an element to the Struct. "All child-builders' Append method 
# must be called independently to maintain data-structure consistency."
# Pass "true" for is_valid.
self.builder.get().Append(True)
```

File: bindings/python/pymongoarrow/lib.pyx [hunk 5]
```diff
@@ -103,72 +147,90 @@ def process_bson_stream(bson_stream, context):
...
+                    else:
+                        builder.append_null()
+                elif ftype == t_bool:
+                    if value_t == BSON_TYPE_BOOL:
+                        builder.append(bson_iter_bool(&doc_iter))
+                    else:
+                        builder.append_null()
+                elif ftype == t_document:
+                    if value_t == BSON_TYPE_DOCUMENT:
+                        bson_iter_document(&doc_iter, &doc_buf_len, &doc_buf)
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:228]: The function returns a void.  I added a guard on the `doc_buf_len`.

File: bindings/python/test/test_builders.py [hunk 6]
```diff
@@ -146,6 +148,39 @@ def test_simple(self):
...
+
+    def test_nested(self):
+        sub_struct = struct([field("c", bool_()), field("d", ObjectIdType())])
+        dtype = struct([field("a", int32()), field("b", sub_struct)])
+        builder = DocumentBuilder(dtype)
+        builder.append({"a": 1, "b": {"c": True, "d": ObjectId()}})
+        builder.append_values(
+            [
+                {"a": 1, "b": {"c": False, "d": ObjectId()}},
+                {"a": 2, "b": {"c": True, "d": ObjectId()}},
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/test/test_builders.py:173]: Done

File: bindings/python/docs/source/changelog.rst [hunk 7]
```diff
@@ -1,6 +1,10 @@
 Changelog
 =========
 
+Changes in Version 0.7.0
+------------------------
+- Added support for BSON Embedded Document type.
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/docs/source/changelog.rst:6]: Done

File: bindings/python/pymongoarrow/lib.pyx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/pymongoarrow/lib.pyx:542]: Sorry I still don't follow this. If we only append True/Null to the builder where are the structs getting saved?

**Review by ShaneHarvey — COMMENTED** [review]
Great work! I have one question on the DocumentBuilder class: https://github.com/mongodb-labs/mongo-arrow/pull/104/files#r1048937009

File: bindings/python/pymongoarrow/lib.pyx (hunk 4 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongoarrow/lib.pyx:542]: This is essentially a marker in the parent builder to say "I've added data to your child builders", which is happening in `process_bson_stream`.

File: bindings/python/test/test_arrow.py [hunk 8]
```diff
@@ -282,28 +287,55 @@ def test_write_batching(self, mock):
...
+            return dict(
+                str=str(i),
+                bool=bool(i),
+                float=i + 0.1,
+                Int64=i,
+                int=i,
+                datetime=datetime(1970 + i, 1, 1),
+            )
+
+        raw_data["nested"] = [inner(i) for i in range(3)]
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/test/test_arrow.py]: Done

**Pull request merged**